lib/types/jtd-schema.ts

```diff
@@ -0,0 +1,103 @@
+/** required keys of an object, not undefined */
+type RequiredKeys<T> = {
+  [K in keyof T]-?: undefined extends T[K] ? never : K
+}[keyof T]
+
+/** optional or undifined-able keys of an object */
+type OptionalKeys<T> = {
+  [K in keyof T]-?: undefined extends T[K] ? K : never
+}[keyof T]
+
+/** type is true if all elements of T extend T */
+type AllExtend_<T, E> = T extends E ? true : false
+type AllExtend<T, E> = false extends AllExtend_<T, E> ? false : true
+
+/** type is true if type is identically string */
+type IsString_<T> = T extends string ? (string extends T ? true : false) : false
+type IsString<T> = false extends IsString_<T> ? false : true
+
+/** gets only the string literals of a type or null if a type isn't a string literal */
+type NullableStringLiterals<T> = T extends string ? (string extends T ? null : T) : null
+
+/** numeric strings */
+type NumberType = "float32" | "float64" | "int8" | "uint8" | "int16" | "uint16" | "int32" | "uint32"
+
+/** string strings */
+type StringType = "string" | "timestamp"
+
+/** actual schema */
+export type JTDSchema<T, D extends Record<string, unknown> = Record<string, never>> = (
+  | // refs - accepts any key in definitions of the right type
+  {[K in keyof D]: T extends D[K] ? {ref: K} : never}[keyof D]
+  // numbers - only accepts number or union of number literals
+  | (AllExtend<T, number | null> extends true ? {type: NumberType} : never)
+  // booleans - accepts any boolean, incluting just true or false literals
+  | (AllExtend<T, boolean | null> extends true ? {type: "boolean"} : never)
+  // strings - only accepts the type string
+  // TODO accept Date?
```

Comment:
yes - Date should be supported and require timestamp type

Reply:
To be clear with the way that AjvJTD works. `JTDSchemaType<Date>` should be `{ type: "timestamp" }`?

Reply:
I think so, yes. Need to test that it works - it only runs standard JTD tests atm.